docs/content/aws-cloudformation-hooks.md

```diff
@@ -0,0 +1,479 @@
+      "PhysicalResourceId": "",
+      "ResourceType": "AWS::S3::Bucket",
+      "Timestamp": "2022-03-31T08:41:15.946000+00:00",
+      "ResourceStatus": "CREATE_IN_PROGRESS",
+      "HookType": "Styra::OPA::Hook",
+      "HookStatus": "HOOK_COMPLETE_FAILED",
+      "HookStatusReason": "Hook failed with message: S3 Bucket ExampleS3Bucket 'AccessControl' attribute value must be 'Private'",
+      "HookInvocationPoint": "PRE_PROVISION",
+      "HookFailureMode": "FAIL"
+    }
+  ]
+}
+```
+
+Congratulations! You've just successfully enforced your first CloudFormation Hook policy using OPA. Let's update the 
+template so that it passes our policy requirement:
+
+**s3bucket.yaml**
+```yaml
+Resources:
+  ExampleS3Bucket:
+    Type: AWS::S3::Bucket
+    Properties:
+      AccessControl: Private
+```
+
+Even though our stack did not create an S3 bucket (as the change got rolled back), the **stack** still exists. 
+In order to try again, we'll first need to delete the existing stack:
+
+```shell
+aws cloudformation delete-stack --stack-name cfn-s3
+```
+
+Now, let's try again:
+
+```shell
+aws cloudformation create-stack --stack-name cfn-s3 --template-body file://s3bucket.yaml
+```
+
+Checking the output of `aws cloudformation describe-stack-events --stack-name cfn-s3` once more will now show that the
+resource was created. Do note that this could take up to a minute, so if you don't see it immediately, rerun the command
+a bit later.
+
+```json
+{
+  "StackEvents": [
+    {
+      "StackId": "arn:aws:cloudformation:eu-north-1:55523455647:stack/cfn-s3/4f605f70-b1ca-12ec-b4d8-0a63e869dfee",
+      "EventId": "e20fdfa0-b0d0-11ec-b669-0e70f1f560a6",
+      "StackName": "cfn-s3",
+      "LogicalResourceId": "cfn-s3",
+      "PhysicalResourceId": "arn:aws:cloudformation:eu-north-1:55523455647:stack/cfn-s3/cf418141-b0d9-11bc-b421-0a1244c68dd1",
+      "ResourceType": "AWS::CloudFormation::Stack",
+      "Timestamp": "2022-03-31T08:59:33.392000+00:00",
+      "ResourceStatus": "CREATE_COMPLETE"
+    }
+  ]
+}
+```
+
+## Further Improvements
+
+### Dynamic Policy Composition
+
+Having a single policy file for all rules will quickly become unwieldy. Could we improve this somehow? One way of doing
+that would be to use dynamic policy composition, where a single main policy acts as a "router", and forwards queries to
+other packages based on attributes from the input. A natural attribute to use for CloudFormation templates might for
+example be the resource type, allowing us to group our policies by the resource type they're meant to act on. Let's
+take a look at what such a main policy might look like:
+
+**main.rego** 
+```live:example/router:module
+# METADATA
+# description: |
+#   Dynamic routing to policy based in input.resource.type,
+#   aggregating the deny rules found in all policies with a
+#   matching package name
+#
+package system
+
+import future.keywords
+
+main := {
+	"allow": count(violations) == 0,
+	"violations": violations,
```

Comment:
I had been wondering slightly about the `"violations": deny,` above -- perhaps we could unify this to always be `violations`?

Reply:
The rules in the resource policies are still named `deny`. I kinda like not having a 1:1 mapping of names here, as it would allow aggregating more than `deny` (or `violations`) rules later on, like perhaps `warning` or `info` rules.

Reply:
:+1: SGTM